libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/NewEpochState.hs

```diff
@@ -0,0 +1,135 @@
+import Test.Cardano.Ledger.Constrained.Monad (monadTyped)
+import Test.Cardano.Ledger.Constrained.Preds.CertState (dstateStage, pstateStage, vstateStage)
+import Test.Cardano.Ledger.Constrained.Preds.LedgerState (ledgerStateStage)
+import Test.Cardano.Ledger.Constrained.Preds.PParams (pParamsStage)
+import Test.Cardano.Ledger.Constrained.Preds.Repl (ReplMode (..), modeRepl)
+import Test.Cardano.Ledger.Constrained.Preds.Universes (universeStage)
+import Test.Cardano.Ledger.Constrained.Rewrite (OrderInfo (..), standardOrderInfo)
+import Test.Cardano.Ledger.Constrained.Solver (toolChainSub)
+import Test.Cardano.Ledger.Constrained.TypeRep
+import Test.Cardano.Ledger.Constrained.Vars
+import Test.Cardano.Ledger.Generic.PrettyCore (pcEpochState, pcNewEpochState)
+import Test.Cardano.Ledger.Generic.Proof
+import Test.QuickCheck
+import Test.Tasty (TestTree, defaultMain)
+
+-- ===========================================================
+
+epochstatePreds :: EraGov era => Proof era -> [Pred era]
+epochstatePreds _proof =
+  [ Dom markStake `Subset` credsUniv
+  , Dom markDelegs `Subset` credsUniv
+  , Rng markDelegs `Subset` poolHashUniv
+  , Sized (Range 1 2) markPools
+  , Dom markPools `Subset` poolHashUniv
+  , Dom setStake `Subset` credsUniv
+  , Dom setDelegs `Subset` credsUniv
+  , Rng setDelegs `Subset` poolHashUniv
+  , Sized (Range 1 2) setPools
+  , Dom setPools `Subset` poolHashUniv
+  , Dom goStake `Subset` credsUniv
+  , Dom goDelegs `Subset` credsUniv
+  , Rng goDelegs `Subset` poolHashUniv
+  , Sized (Range 1 2) goPools
+  , Dom goPools `Subset` poolHashUniv
+  , Random snapShotFee
+  , SumsTo (Right (1 % 1000)) (Lit RationalR 1) EQL [ProjMap RationalR individualPoolStakeL markPoolDistr]
+  , Sized (Range 2 6) markPoolDistr
+  ]
+
+newepochstatePreds :: Era era => Proof era -> [Pred era]
+newepochstatePreds _proof =
+  [ Sized (ExactSize 4) (Dom prevBlocksMade) -- Both prevBlocksMade and prevBlocksMadeDom will have size 4
+  , Sized (ExactSize 4) (Dom currBlocksMade)
+  , Dom prevBlocksMade `Subset` poolHashUniv
+  , Dom currBlocksMade `Subset` poolHashUniv
+  ]
+
+-- ========================
+
+epochStateStage ::
+  Reflect era =>
+  Proof era ->
+  Subst era ->
+  Gen (Subst era)
+epochStateStage proof = toolChainSub proof standardOrderInfo (epochstatePreds proof)
+
+demoES :: Reflect era => Proof era -> ReplMode -> IO ()
+demoES proof mode = do
+  env <-
+    generate
+      ( pure emptySubst
+          >>= pParamsStage proof
+          >>= universeStage def proof
+          >>= vstateStage proof
+          >>= pstateStage proof
+          >>= dstateStage proof
+          >>= ledgerStateStage def proof
+          >>= epochStateStage proof
+          >>= (\subst -> monadTyped $ substToEnv subst emptyEnv)
+      )
+  epochstate <- monadTyped $ runTarget env (epochStateT proof)
+  let env2 = getTarget epochstate (epochStateT proof) emptyEnv
+  when (mode == Interactive) $ putStrLn (show (pcEpochState proof epochstate))
```

Comment:
```suggestion
  when (mode == Interactive) $ print (pcEpochState proof epochstate)
```

Seems like a pattern that is used a lot, maybe adding a helper function for it?
```haskell
printInteractive mode = when (mode == Interactive) . print
```

Reply:
done